crank/tests/test_stack_api.py

```diff
@@ -0,0 +1,192 @@
+            else:
+                print("Finished")
+                return self.collect_lowest_energies(in_dict)
+
+    def create_initial_api_input(self):
+        """ Create the initial input dictionary for crank-api """
+        return {
+            'dihedrals': self.dihedrals,
+            'grid_spacing': self.grid_spacing,
+            'elements': self.M.elem,
+            'init_coords': [self.M.xyzs[0].ravel().tolist()],
+            'grid_status': {}
+        }
+
+    def run_jobs(self, next_jobs):
+        """ Take a dictionary of next jobs, run them and return the results """
+        job_results = collections.defaultdict(list)
+        for grid_id_str, job_geo_list in next_jobs.items():
+            grid_id = tuple(int(i) for i in grid_id_str.split(','))
+            dihedral_idx_values = []
+            for dihedral_idxs, dihedral_value in zip(self.dihedrals, grid_id):
+                dihedral_idx_values.append(list(dihedral_idxs) + [dihedral_value])
+            # running jobs in serial here, if run in parallel the order should be kept same
+            for job_geo in job_geo_list:
+                # run the job using engine
+                self.engine.M.xyzs = [np.array(job_geo, dtype=float).reshape(-1, 3)]
+                self.engine.set_dihedral_constraints(dihedral_idx_values)
+                self.engine.optimize_geomeTRIC()
+                m = self.engine.load_geomeTRIC_output()
+                # here we check if the dihedrals after optimization matches the constraint
+                # this is good for validity if the engine is not checking that
+                if "final dihedral may be different than constraint":
+                    dihedral_values = np.array([m.measure_dihedrals(*d)[0] for d in self.dihedrals])
+                    for dv, dref in zip(dihedral_values, grid_id):
+                        diff = abs(dv - dref)
+                        if min(diff, abs(360-diff)) > 0.9:
+                            print("Warning! dihedral values inconsistent with check_grid_id")
+                            print('dihedral_values', dihedral_values, 'ref_grid_id', grid_id)
+                    dihedral_id = (np.round(dihedral_values / self.grid_spacing) * self.grid_spacing).astype(int)
+                    result_grid_id = [(d + (180-d)//360*360) for d in dihedral_id]
+                else:
+                    result_grid_id = grid_id
+                # put the results into job_results dict, note that the order is same as input
+                result_grid_id_str = ','.join(map(str, result_grid_id))
+                start_geo = job_geo
+                end_geo = m.xyzs[0].ravel().tolist()
+                end_energy = m.qm_energies[0]
+                job_result_tuple = (start_geo, end_geo, end_energy)
+                job_results[result_grid_id_str].append(job_result_tuple)
+        return job_results
+
+    def update_api_input(self, in_dict, job_results):
```

Comment:
This and the function below should be moved into the `crank_api` calls so that user has a much simpler objective.

```
data = start_crank_data
program = "psi4"
for x in range(max_iter):
    new_compute = crank_api(data)
    if new_compute == "finished": # or whatever
        break 
    compute = {}
    for key, job in new_compute.items():
          job = process_geometry(..., job) # Or maybe even return full JSON jobs so no process geometry is needed.
          compute[key] = geometric.run_json(job, program)
    data = crank_update_data(data, compute)

lowest_energy = crank_process_lowest_energies(data)
```

Where the `crank_*` functions come from Crank itself.

Reply:
Similar to above, I agree that it would be nice if the server just calls functions from different modules and watch them work together nicely, however I think it is also an important feature for the server to be able to interface with difference APIs, which means it will parse the outputs and generate inputs. These functionalities don't need to be implemented in the server, they can still be imported and called by the server.

Reply:
Maybe I should put the structure of the class a little differently in this test. Many functions can be seperated out from the `FakeServer` and put into another class, like `CrankServerAPI`, so the real server can import that `CrankServerAPI` and call these functions there.